src/app/loadout-builder/updated-loadout.ts

```diff
@@ -0,0 +1,92 @@
+import { DimItem, PluggableInventoryItemDefinition } from 'app/inventory/item-types';
+import { Loadout } from 'app/loadout-drawer/loadout-types';
+import { convertToLoadoutItem } from 'app/loadout-drawer/loadout-utils';
+import { BucketHashes } from 'data/d2/generated-enums';
+import { t } from 'i18next';
+import _ from 'lodash';
+import { ArmorSet, LockableBucketHashes } from './types';
+import { statTier } from './utils';
+
+/**
+ * Create a new loadout from the original prototype loadout, but with the armor
+ * items replaced with this loadout's armor. Used for equipping or creating a
+ * new saved loadout.
+ */
+export function updateLoadoutWithArmorSet(
+  defs: D2ManifestDefinitions,
+  loadout: Loadout,
+  set: ArmorSet,
+  items: DimItem[],
+  lockedMods: PluggableInventoryItemDefinition[]
+): Loadout {
+  const data = {
+    tier: _.sumBy(Object.values(set.stats), statTier),
+  };
+
+  const existingItemsWithoutArmor = loadout.items.filter(
+    (li) =>
+      // The new item might already be in the loadout (but unequipped), remove it
+      !items.some((i) => i.id === li.id) &&
+      // Remove equipped armor items
+      !(
+        li.equip &&
+        LockableBucketHashes.includes(
+          defs.InventoryItem.get(li.hash)?.inventory?.bucketTypeHash ?? 0
+        )
+      )
+  );
+  const loadoutItems = items.map((item) => convertToLoadoutItem(item, true));
+
+  // We need to add in this set's specific stat mods (artifice, general) to the
+  // list of user-chosen mods We can't start with the list of mods in the
+  // existing loadout parameters because lockedMods has filtered out invalid
+  // mods, mods that don't fit, and general mods if we're auto-assigning general
+  // mods.
+  const allMods = [...lockedMods.map((m) => m.hash), ...set.statMods];
+  return {
+    ...loadout,
+    parameters: {
+      ...loadout.parameters,
+      mods: allMods.length ? allMods : undefined,
+    },
+    items: [...existingItemsWithoutArmor, ...loadoutItems],
+    name: loadout.name ?? t('Loadouts.Generated', data),
+  };
+}
+
+/**
+ * Create a new loadout from an original prototype loadout, using mods and
+ * subclass from another loadout, and the items from an armor set. Used for the
+ * "compare loadout" drawer.
+ */
+export function mergeLoadout(
+  defs: D2ManifestDefinitions,
+  originalLoadout: Loadout,
+  newLoadout: Loadout,
+  set: ArmorSet,
+  items: DimItem[],
+  lockedMods: PluggableInventoryItemDefinition[]
+): Loadout {
+  const loadoutWithArmorSet = updateLoadoutWithArmorSet(
+    defs,
+    originalLoadout,
+    set,
+    items,
+    lockedMods
+  );
+
+  const newSubclass = newLoadout.items.find(
+    (li) => defs.InventoryItem.get(li.hash)?.inventory?.bucketTypeHash === BucketHashes.Subclass
+  );
+
+  if (newSubclass) {
+    const itemsWithoutSubclass = loadoutWithArmorSet.items.filter(
+      (li) => defs.InventoryItem.get(li.hash)?.inventory?.bucketTypeHash !== BucketHashes.Subclass
+    );
```

Comment:
Strictly speaking, accessing loadout items by hash isn't 100% safe since we have migration for subclasses with `oldToNewHash` (I see a similar issue in `loadout-builder-reducer:getBucketHashFromItemHash`) but as long as the old defs keep existing and says they're subclasses I think it's fine?

Reply:
Yeah, overall it'd be more correct to operate on resolved items always, but that introduces a lot of questions around how we'd structure reducer-style state management.